### PR TITLE
Use the Koa context rather than handling dependencies separately

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Current the router is being passed around separately, rather than using the Koa context. This adopts the pattern used by the framework (and is already implemented in the `@koa/router` library).

I've kept the factory pattern in place, not sure whether to remove it where possible.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,3 +1,24 @@
...
+
+app.use(logger());
+app.use(cors({
+  exposeHeaders: ['Link'],
+}));
+app.use(apiDocumentation());
+app.use(errorHandler());
+app.use(routing(app.context.router));
+
+export default app.listen(8080);
```
Comment (thewilkybarkid) [inline src/index.ts]: This change should be extracted.

Comment (giorgiosironi) [reply src/index.ts]: https://github.com/libero/article-store/pull/80 approved and will make this PR leaner.

Comment (giorgiosironi) [reply src/index.ts]: Actually doesn't look like the same thing, but will surely conflict?

Comment (thewilkybarkid) [reply src/index.ts]: Just pushed https://github.com/libero/article-store/pull/60/commits/6be8217544419a4127648caf4f6abda4bc96a82a which updates this PR to include #80.

**Comment by thewilkybarkid** [comment]
Really wanted `types.ts` to be in `app.ts`, but end up with circular references.

File: src/types.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+export type AppState = DefaultState;
+
+export type AppContext = RouterContext<AppState, {
+  articles: Articles;
+}>;
+
+export type AppMiddleware = Middleware<AppState, AppContext>;
+
+export type App = Koa<AppState, AppContext>;
```
Comment (thewilkybarkid) [inline src/types.ts]: Did want these inside `app.ts`, but end up with circular references otherwise.

Comment (stephenwf) [reply src/types.ts]: Is `App` required? It should be inferred correctly from the default export returning in `app.ts`

Comment (thewilkybarkid) [reply src/types.ts]: Looking at whether it can be reduced to just `AppContext`, as that's the important one.

Comment (thewilkybarkid) [reply src/types.ts]: Nope. The separate router is the problem...

Comment (thewilkybarkid) [reply src/types.ts]: In fact, even moving `createRouter` doesn't help (since the routes themselves are now circular).

Comment (thewilkybarkid) [reply src/types.ts]: Might have to inject the router into the app factory? Shame, as the `apiDocumentation` middleware relies on a specific route. Might have to inject that too?

(Feels like testing will be a bit weirder.)

Comment (stephenwf) [reply src/types.ts]: You could avoid that by making the route generic? 

```typescript
function createRouter<State, Context>() {
  const router = new Router<State, Context>();

  router.get( 
     //... 
  );
}
```
Although might need to narrow the types on the generics

Comment (thewilkybarkid) [reply src/types.ts]: Tried that, not assignable in https://github.com/libero/article-store/blob/1acc2dadafdd154f9b6f5f61fb9545524da29be3/src/app.ts#L17 😞

Comment (thewilkybarkid) [reply src/types.ts]: Thinking about making the path an argument on `apiDocumentation`, so it's decoupled from the router.

File: src/routes/article-list.ts [hunk 3]
```diff
@@ -1,11 +1,12 @@
...
 import { hydra, rdf, schema } from 'rdf-namespaces';
-import Articles from '../articles';
+import { AppContext, AppMiddleware } from '../types';
 import Routes from './index';
 
-export default (articles: Articles, router: Router): Middleware => (
-  async ({ request, response }: Context, next: Next): Promise<void> => {
+export default (): AppMiddleware => (
+  async ({
+    articles, request, response, router,
```
Comment (giorgiosironi) [inline src/routes/article-list.ts:8]: it is a refactoring from Dependency Injection to a Service Locator (the `AppContext`), but at least this unpacking makes the dependencies clear

File: src/types.ts (hunk 2 above)
Comment (thewilkybarkid) [reply src/types.ts]: https://github.com/libero/article-store/pull/60/commits/bf78b85d521ce96af83994c6611e7a783b795a60, but needs extracting

Comment (stephenwf) [reply src/types.ts]: Right, I see. Pulled it down to check and yeh it's like the app and router are pretty must one and the same conceptually. I definitely see the benefit in splitting out the routes though.

In this case though, it does make sense to have some types grouped together as a common concept that they are all a part of (the "app").

Comment (thewilkybarkid) [reply src/types.ts]: Push https://github.com/libero/article-store/pull/60/commits/14962894eb80dce81cceb5f12cba8fc23c8a479e which is now possible. Doesn't feel great though.

File: src/app.ts [hunk 4]
```diff
@@ -1,16 +1,30 @@
...
 import errorHandler from './middleware/error-handler';
 import routing from './middleware/routing';
-import createRouter from './router';
 
-export default (articles: Articles): Koa => {
-  const app = new Koa();
-  const router = createRouter(articles);
+export type AppState = DefaultState;
+
+export type AppContext = RouterContext<AppState, {
```
Comment (giorgiosironi) [inline src/app.ts:13]: seems like a good alias

Comment (giorgiosironi) [reply src/app.ts:13]: does it matter that types are inside a top level module like `app`? Though they are used in conjunction with this module so it's probably the right place even if requiring this brings in everything.

File: src/index.ts [hunk 5]
```diff
@@ -1,8 +1,12 @@
...
 import createApp from './app';
+import createRouter from './router';
+import Routes from './routes';
 
 const articles = new InMemoryArticles();
+const router = createRouter();
+const apiDocumentationPath = router.url(Routes.ApiDocumentation);
 
-const app = createApp(articles);
+const app = createApp(articles, router, apiDocumentationPath);
```
Comment (giorgiosironi) [inline src/index.ts:10]: doesn't look like it can scale to lots of dependencies?

File: test/router.test.ts [hunk 6]
```diff
@@ -1,10 +1,9 @@
 import Layer from '@koa/router/lib/layer';
-import InMemoryArticles from '../src/adaptors/in-memory-articles';
 import createRouter from '../src/router';
 import Routes from '../src/routes';
 
 describe('router', (): void => {
-  const router = createRouter(new InMemoryArticles());
+  const router = createRouter();
```
Comment (giorgiosironi) [inline test/router.test.ts:6]: sounds good that to create a router you don't need an `Articles` implementation :thinking:

File: test/routes/entry-point.test.ts [hunk 7]
```diff
@@ -4,11 +4,9 @@ import entryPoint from '../../src/routes/entry-point';
 import runMiddleware from '../middleware';
 import createContext from '../context';
 
-const makeRequest = async (next?: Next): Promise<Response> => {
-  const context = createContext();
-
-  return runMiddleware(entryPoint(context.router), context, next);
-};
+const makeRequest = async (next?: Next): Promise<Response> => (
+  runMiddleware(entryPoint(), createContext(), next)
```
Comment (giorgiosironi) [inline test/routes/entry-point.test.ts:8]: `makeRequest` doesn't return but can still `await` on it to get a response?

Comment (thewilkybarkid) [reply test/routes/entry-point.test.ts:8]: Parentheses rather than curly brackets (line above) are a return.

Comment (giorgiosironi) [reply test/routes/entry-point.test.ts:8]: ah, it's Ruby all over again :+1:

**Review by giorgiosironi — APPROVED** [review]
Long but mostly consistent across updating routes and tests

**Comment by thewilkybarkid** [comment]
Will do https://github.com/libero/article-store/pull/60#discussion_r354792149 before merging this.

File: src/index.ts (hunk 5 above)
Comment (thewilkybarkid) [reply src/index.ts:10]: Could remove `apiDocumentationPath`, and assume the router has a route named `Routes.ApiDocumentation`...

**Pull request merged**